backend/data_adapter.go

```diff
@@ -45,8 +45,8 @@ func (a *dataSDKAdapter) QueryData(ctx context.Context, req *pluginv2.QueryDataR
 				continue
 			}
 
-			// if error source not set and the error is a downstream error, set error source to downstream.
-			if !r.ErrorSource.IsValid() && IsDownstreamError(r.Error) {
+			// we check if error is downstream error and set the error source to downstream.
+			if IsDownstreamError(r.Error) {
```

Comment:
Should this rather be:
```suggestion
			if !r.ErrorSource.IsValid() || IsDownstreamError(r.Error) {
```
🤔

Reply:
Hmm, depends. Do we care about if the error source is empty, has downstream error or plugin error when evaluating this? 

Because if we went with your change, this test would fail, but I would expect it in reality to pass. 
```
			{
				name: `single error that we override as "downstream" should be "downstream" error source even if no error source`,
				queryDataResponse: &QueryDataResponse{
					Responses: map[string]DataResponse{
						"A": {Error: context.Canceled},
					},
				},
				expErrorSource: ErrorSourceDownstream,
			},
```

Reply:
That test would be fine but I see this one would fail:

```
{
				name: `multiple downstream error without error source and single plugin error should be "plugin" error source`,
				queryDataResponse: &QueryDataResponse{
					Responses: map[string]DataResponse{
						"A": {Error: DownstreamErrorf("boom")},
						"B": {Error: someErr},
						"C": {Error: DownstreamErrorf("boom")},
					},
				},
				expErrorSource: ErrorSourcePlugin,
			},
```

I'm just trying to think do we still want a check for `!r.ErrorSource.IsValid()` here somewhere. Could we use it as an opportunity to default to `ErrorSourcePlugin` which could be used to drive to migrate more to error source? Might be blowing things wide open though 😬

Reply:
Something like this? 		
```
// we check if error is downstream error and set the error source to downstream.
			if IsDownstreamError(r.Error) {
				r.ErrorSource = ErrorSourceDownstream
			} else if !r.ErrorSource.IsValid() {
				r.ErrorSource = ErrorSourcePlugin
			}
```

Reply:
Yeah I think so!

Reply:
https://github.com/grafana/grafana-plugin-sdk-go/pull/1097/commits/d25ccce05b3aad0125560247be4dd88f0831c79e

Reply:
The idea with checking `!r.ErrorSource.IsValid()` first is so that we only override if not already set by the plugin author.